### PR TITLE
RoR Capstone project

### PR DESCRIPTION
**Pull request by emmanuellmhango**
#### Design
- You should follow these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
  - Colors.
  - Typography: font face, size and weight.
  - Layout: composition and space between elements.

- Splash screen
  - A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

- Sign up and log in pages
  - The user should be able to register in the app with full name, email and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (categories page)
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new transaction" page
  - The user fills out a form to create a new transaction with:
    - name (mandatory)
    - amount (mandatory)
    - categories (mandatory at least one)
  - The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,188 @@
...
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+
+## 🚀 Live Demo <a name="live-demo"></a>
+
+To see the app, please click the link [LIVE DEMO](https://budget-tracker-w9a5.onrender.com/)
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
```
Comment (omar-muhamad) [inline README.md:69]: - [ ] This [Video Presentation](https://www.loom.com/share/a58a98aa1817470ea1e5395c2f380eb0) you provided suppose to highlight some interesting piece of code or something you built that you are very proud of. As mentioned in this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=you%20should%20also%20highlight%20some%20interesting%20piece%20of%20code%20or%20something%20you%20built%20that%20you%20are%20very%20proud%20of.), Please fix that and kindly add the video link here in the readme file or in the Pull Request summary.

File: app/views/categories/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    <p style="color: green"><%= notice %></p>
+  </center>
+  <br />
+  <% @categories.each do |category| %>
+    <%= render category %>
+  <% end %>
+  <br />
+  <center>
+    <%= link_to "add a new category", new_category_path, class: 'down-btn' %>
+  </center>
```
Comment (omar-muhamad) [inline app/views/categories/index.html.erb]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=size%20and%20weight.-,Layout%3A%20composition%20and%20space%20between%20elements.,-NOTE%3A%20In%20these), Please add the correct placement for this button at the end of the page and make it take the whole page width to match the given design.

| Given Design  | Your Design |
| ------------- | ------------- |
| image  | image  |

File: app/views/categories/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  </div>
+</div>
+<div class="categories">
+  <center>
+    <p style="color: green"><%= notice %></p>
+  </center>
+  <br />
+  <% @categories.each do |category| %>
+    <%= render category %>
+  <% end %>
```
Comment (omar-muhamad) [inline app/views/categories/index.html.erb:31]: - [ ] Kindly fix the category card elements placement and add the correct widths and margin values to match the given design.

| Given Design  | Your Design |
| ------------- | ------------- |
| image | image  |

File: app/views/purchases/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+</div>
+
+<div class="categories">
+  <center>
+    <p style="color: green"><%= notice %></p>
+  </center>
+  <br />
+  <% @purchases.each do |purchase| %>
+    <%= render purchase %>
+  <% end %>
```
Comment (omar-muhamad) [inline app/views/purchases/index.html.erb:38]: - [ ] same here,  Kindly fix the transaction card elements placement and add the correct widths and margin values to match the given design.

| Given Design  | Your Design |
| ------------- | ------------- |
|image | image |

File: app/views/purchases/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  <center>
+    <%= link_to "New purchase", new_purchase_path, class: 'down-btn' %>
+  </center>
```
Comment (omar-muhamad) [inline app/views/purchases/index.html.erb]: - [ ] Same here, As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=size%20and%20weight.-,Layout%3A%20composition%20and%20space%20between%20elements.,-NOTE%3A%20In%20these), Please add the correct placement for this button at the end of the page and make it take the whole page width to match the given design.

| Given Design  | Your Design |
| ------------- | ------------- |
| image  | image |

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @emmanuellmhango,

### Nice job, you are almost there.

cat-driving-serious

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+class ApplicationController < ActionController::Base
+  before_action :authenticate_user!
+  protect_from_forgery with: :exception
+
+  before_action :update_allowed_parameters, if: :devise_controller?
+
+  def after_sign_in_path_for(_resource)
```
Comment (arslanbisharat) [inline app/controllers/application_controller.rb:7]: - [ ] Upon signning up, I am getting error. Please fix this. 

image

File: app/views/categories/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<h1>Editing category</h1>
+
```
Comment (arslanbisharat) [inline app/views/categories/edit.html.erb:2]: - [ ] After signup and getting an error, If I try to again sign up I get an error of email already being used. But if I try to use the same Email and Password in Login page, it is not working. It also doesn't have any error.

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
Hi @emmanuellmhango, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

- The Pull Request has a proper title and description. ✔️
- You have added a descriptive Readme file. Good Job. ✔️
- There is no linter error is present. ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [ ] Please make sure you right descriptive comments. Commenting involves placing Human Readable Descriptions inside of computer programs detailing what the Code is doing. Proper use of commenting can make code maintenance much easier, as well as helping make finding bugs faster. Further, commenting is very important when writing functions that other people will use. :x:


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by hiromataba — APPROVED** [review]
# Status: Approved :tada: :tada: :tada: 
Hi @emmanuellmhango    :wave: :wave: 

200w  

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**